build/helper/metadata_add_all.py

```diff
@@ -471,8 +486,16 @@ def _add_enum_codegen_method(enums, config):
             elif e is not None:
                 if a_codegen_method == 'private' and enums[e]['codegen_method'] == 'no':
                     enums[e]['codegen_method'] = a_codegen_method
-                elif a_codegen_method == 'public':
-                    enums[e]['codegen_method'] = a_codegen_method
+                elif a_codegen_method == 'public' and enums[e]['codegen_method'] != 'public':
+                    if enum_to_has_explicit_codegen_method[e]:
+                        if not enum_uses_converter(enums[e]):
+                            raise ValueError(
+                                "Public attribute {} cannot use non-public enum {} (without converter)".format(
+                                    config['attributes'][a]['name'], e
+                                )
+                            )
+                    else:
+                        enums[e]['codegen_method'] = a_codegen_method
```

Comment:
I haven't really found a nice way to refactor it as some parts of the function are similar but not exactly the same, e.g. the wording of 'attribute' vs. 'function'. Nevertheless, I did discover that the `if f_codegen_method != 'no':` and `if a_codegen_method != 'no':` checks were redundant since `filter_codegen_functions()` and `filter_codegen_attributes()` had already performed these checks, so I removed them from the code.